File: Cpp/0486_predict_the_winner.cpp
// https://leetcode.cn/problems/predict-the-winner/

/* Check solution for a DP-based method. */

// class Solution {
// public:
//     bool PredictTheWinner(vector<int>& nums) {
//         nums_ = nums;
//         if (dfs(0, nums.size()-1, 0, 0) >= 0) {
//             return true;
//         }
//         else {
//             return false;
//         }
//     }

//     int dfs(int left, int right, int total1, int total2) {
//         if (left + 1 >= right) {
//             total1 += max(nums_[left], nums_[right]);
//             total2 += min(nums_[left], nums_[right]);
//             return total1 - total2;
//         }

//         int diff1 = -dfs(left + 1, right, total2, total1 + nums_[left]);
//         int diff2 = -dfs(left, right - 1, total2, total1 + nums_[right]);
//         return max(diff1, diff2);
//     }

//     vector<int> nums_;
// };


// This is an improved version based on the idea above with repeated operations avoided.
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        nums_ = nums;
        n_ = nums.size();
        if (dfs(0, n_-1) >= 0) {
            return true;
        }
        else {
            return false;
        }
    }

    int dfs(int left, int right) {
        int key = left * n_ + right;
        if (records_.find(key) != records_.end()) {
            return records_[key];
        }
        if (left + 1 >= right) {
            int diff = abs(nums_[left] - nums_[right]);
            records_[key] = diff;
            return diff;
        }

        int diff1 = nums_[left] - dfs(left + 1, right);
        int diff2 = nums_[right] - dfs(left, right - 1);
        int diff = max(diff1, diff2);
        records_[key] = diff;
        return diff;
    }

    vector<int> nums_;
    int n_;
    unordered_map<int, int> records_;
};

File: Python3/0486_predict_the_winner.py
# https://leetcode.cn/problems/predict-the-winner/

"""
Check solution for a DP-based method.
"""

# class Solution:
#     def PredictTheWinner(self, nums):
#         """
#         Inputs:
#             nums: list[int]
#         Outputs:
#             res: bool
#         """
#         self.nums = nums
#         if self.dfs(0, len(nums)-1, 0, 0) >= 0:
#             return True
#         else:
#             return False

#     def dfs(self, left, right, total1, total2):
#         if left + 1 >= right:
#             total1 += max(self.nums[left], self.nums[right])
#             total2 += min(self.nums[left], self.nums[right])
#             return total1 - total2
        
#         diff1 = -self.dfs(left+1, right, total2, total1 + self.nums[left])
#         diff2 = -self.dfs(left, right-1, total2, total1 + self.nums[right])
#         return max(diff1, diff2)


# This is an improved version based on the idea above with repeated operations avoided.
class Solution:
    def PredictTheWinner(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: bool
        """
        self.nums = nums
        self.n = len(nums)
        self.records = dict()
        if self.dfs(0, self.n-1) >= 0:
            return True
        else:
            return False

    def dfs(self, left, right):
        key = left * self.n + right
        if key in self.records:
            return self.records[key]
        if left + 1 >= right:
            diff = abs(self.nums[left] - self.nums[right])
            self.records[key] = diff
            return diff
        
        diff1 = self.nums[left] - self.dfs(left+1, right)
        diff2 = self.nums[right] - self.dfs(left, right-1)
        diff = max(diff1, diff2)
        self.records[key] = diff
        return diff
